### PR TITLE
Protected monolog

### LINKED ISSUE #5366: Conflict between Monolog 1.x and 2.x
**Describe the bug**
In a recent update (version 3.12.0.3), WP Rocket seems to have started using version 2.x of the Monolog library. This causes conflicts with plugins that still use version 1.x (all of mine, for example). 

Although the interface of the two versions is compatible, the code in itself is not. Due to that, when the autoloader fetches some files from version 1.x and some from version 2.x, that triggers a fatal error. Example:
```
Fatal error: Uncaught Error: Undefined class constant 'Monolog\Utils::DEFAULT_JSON_FLAGS' in /path/to/plugins/wc-aelia-foundation-classes/src/vendor/monolog/monolog/src/Monolog/Handler/AbstractHandler.php on line 194
Call stack:
#0 Monolog\H\AbstractHandler::getDefaultFormatter() wp-content/plugins/wc-aelia-foundation-classes/src/vendor/monolog/monolog/src/Monolog/Handler/AbstractHandler.php:113
#1 Monolog\H\AbstractHandler::getFormatter() wp-content/plugins/wc-aelia-foundation-classes/src/vendor/monolog/monolog/src/Monolog/Handler/AbstractProcessingHandler.php:37
#2 Monolog\H\AbstractProcessingHandler::handle() wp-content/plugins/wc-aelia-foundation-classes/src/vendor/monolog/monolog/src/Monolog/Logger.php:344
#3 Monolog\Logger::addRecord() wp-content/plugins/wc-aelia-foundation-classes/src/vendor/monolog/monolog/src/Monolog/Logger.php:642
```
The error refers of files loaded from Monolog 1.x, located at `/wc-aelia-foundation-classes/src/vendor/monolog/monolog`. For some reason, when the class `AbstractHandler` loads, it loads a line formatter from Monolog 2.x, such as `NormalizerFormatter` or `HtmlFormatter`. These reference constant `Utils::DEFAULT_JSON_FLAGS`, which doesn't exist in Monolog 1.x, thus causing a fatal error.

**To Reproduce**
1. Install and activate WP Rocket with Monolog 2.x.
2. Install and activate a plugin that runs Monolog 1.x. Our Aelia Foundation Classes and the Currency Switcher (the latter is not public, but I can provide a copy of it).
3. Open a page on the frontend that triggers the logging of some operation. In the case of the Currency Switcher, a simple currency selection using the widget can do the trick.

**Expected behavior**
There shouldn't be a fatal error caused by conflicting versions of the Monolog library.  This could be achieved in two ways:

1. Rolling back the Monolog library in WP Rocket to 1.x. The older version works like 2.x, in terms of calls, therefore you wouldn't have to make changes to your code.
2. Namespacing the Monolog library used WP Rocket, with a tool like [Mozart](https://github.com/coenjacobs/mozart), so that it doesn't conflict with the root namespace. This would allow you to keep Monolog 2.x in your code, without affecting any other plugin that uses version 1.x.

**Additional context**
The issue occurs whenever pieces of two incompatible versions of the Monolog library are loaded at the same time. We can't move to version 2.x because we must maintain compatibility with PHP 7.1 for a while longer. Also, there are other plugins that could be using Monolog 1.x, for the same reason, and upgrading to 2.x would simply "shift" the conflict onto them.

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description

Protected the monolog package by creating and importing a custom version and using Mozart on it.

Fixes #5366

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which improves an existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Is the solution different from the one proposed during the grooming?

No.

## How Has This Been Tested?

- [ ] Tested locally

# Checklist:

Please delete the options that are not relevant.

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/Engine/Cache/PurgeActionsSubscriber.php [hunk 1]
```diff
@@ -57,6 +57,7 @@ public static function get_subscribed_events() {
 			'rocket_after_save_dynamic_lists'     => 'purge_cache',
 			'update_option_' . $slug              => [ 'purge_cache_reject_uri_partially', 10, 2 ],
 			'update_option_blog_public'           => 'purge_cache',
+			'wp_rocket_upgrade'                   => [ 'on_update', 10, 2 ],
```
Comment (jeawhanlee) [inline inc/Engine/Cache/PurgeActionsSubscriber.php]: I think we move this hook and it's callback to the `WP_Rocket\Engine\Cache::AdminSubscriber` instead.
What do you think?

File: inc/Logger/HTMLFormatter.php [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <?php
 namespace WP_Rocket\Logger;
```
Comment (jeawhanlee) [inline inc/Logger/HTMLFormatter.php:2]: Do you think we need to move the Logger classes out of the `classes` directory?

Comment (CrochetFeve0251) [reply inc/Logger/HTMLFormatter.php:2]: @jeawhanlee we are touching the `advanced-cache.php` file that's the moment or never to refactor that part.

**Review by jeawhanlee — APPROVED** [review]

**Review by jeawhanlee — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@CrochetFeve0251 Thanks for the PR. 
We couldn't reproduce the initial issue using newlabs test env, were you able to reproduce it on any of our test sites?
However, when installing the branch using cmd `composer install --no-dev --no-scripts`, then accessing the dashboard, can see this error.

`[30-May-2023 06:26:29 UTC] PHP Fatal error:  Uncaught Error: Class "WP_Rocket\Plugin" not found in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/main.php:46
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(308): rocket_init()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(517): WP_Hook->do_action()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-settings.php(486): do_action()
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-config.php(142): require_once('...')
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-load.php(50): require_once('...')
#6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/admin-ajax.php(22): require_once('...')
#7 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/main.php on line 46
[30-May-2023 06:26:58 UTC] PHP Fatal error:  Uncaught Error: Class "WP_Rocket\Logger\Logger" not found in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php:140
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php(108): WP_Rocket\Buffer\Abstract_Buffer->get_process_id()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php(69): WP_Rocket\Buffer\Abstract_Buffer->log()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/advanced-cache.php(75): WP_Rocket\Buffer\Cache->__construct()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-settings.php(95): include('...')
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-config.php(142): require_once('...')
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-load.php(50): require_once('...')
#6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/admin-ajax.php(22): require_once('...')
#7 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php on line 140`

Note:
- Tested on newlabs, cloned repo, switched to the branch then merged dev to branch before the install cmd

**Comment by CrochetFeve0251** [comment]
> @CrochetFeve0251 Thanks for the PR. We couldn't reproduce the initial issue using newlabs test env, were you able to reproduce it on any of our test sites? However, when installing the branch using cmd `composer install --no-dev --no-scripts`, then accessing the dashboard, can see this error.
> 
> `[30-May-2023 06:26:29 UTC] PHP Fatal error: Uncaught Error: Class "WP_Rocket\Plugin" not found in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/main.php:46 Stack trace: #0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(308): rocket_init() #1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters() #2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(517): WP_Hook->do_action() #3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-settings.php(486): do_action() #4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-config.php(142): require_once('...') #5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-load.php(50): require_once('...') #6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/admin-ajax.php(22): require_once('...') #7 {main} thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/main.php on line 46 [30-May-2023 06:26:58 UTC] PHP Fatal error: Uncaught Error: Class "WP_Rocket\Logger\Logger" not found in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php:140 Stack trace: #0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php(108): WP_Rocket\Buffer\Abstract_Buffer->get_process_id() #1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php(69): WP_Rocket\Buffer\Abstract_Buffer->log() #2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/advanced-cache.php(75): WP_Rocket\Buffer\Cache->__construct() #3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-settings.php(95): include('...') #4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-config.php(142): require_once('...') #5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-load.php(50): require_once('...') #6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/admin-ajax.php(22): require_once('...') #7 {main} thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php on line 140`
> 
> Note:
> 
>     * Tested on newlabs, cloned repo, switched to the branch then merged dev to branch before the install cmd

To reproduce the issue you need to install a plugin that use a different version of Monolog.

Concerning the PR this one need to be installed via the plugin update process to execute code to change the `advanced-cache.php`.

**Comment by Mai-Saad** [comment]
@CrochetFeve0251 For fresh install (using zip or cmd) with [event ticket](https://wordpress.org/plugins/event-tickets/) plugin => working fine.
However, if we updated using zip from 3.13.3 to the PR zip while [event ticket](https://wordpress.org/plugins/event-tickets/) plugin is enabled, we will have the following error
`[05-Jun-2023 11:31:07 UTC] PHP Fatal error:  Uncaught Error: Class "WP_Rocket\Logger\Logger" not found in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php:140
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php(108): WP_Rocket\Buffer\Abstract_Buffer->get_process_id()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-cache.php(69): WP_Rocket\Buffer\Abstract_Buffer->log()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/advanced-cache.php(69): WP_Rocket\Buffer\Cache->__construct()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-settings.php(95): include('...')
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-config.php(142): require_once('...')
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-load.php(50): require_once('...')
#6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-login.php(12): require('...')
#7 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/classes/Buffer/class-abstract-buffer.php on line 140`
can you please check :pray: 
[screen-capture.webm](https://github.com/wp-media/wp-rocket/assets/76941962/ef3e28d6-16a8-421f-b403-c2f2d4c3d9c9)

**Comment by CrochetFeve0251** [comment]
@Mai-Saad I added a way to make the new code work with the old `advanced-cache.php` making the update possible

File: composer.json [hunk 3]
```diff
@@ -73,7 +73,8 @@
 		"wpackagist-plugin/simple-custom-css": "^4.0.3",
 		"wpackagist-plugin/spinupwp": "^1.1",
 		"wpackagist-plugin/woocommerce": "^7",
-		"wpackagist-plugin/wp-smushit": "^3"
+		"wpackagist-plugin/wp-smushit": "^3",
+		"wp-media/monolog": "^0.0"
```
Comment (remyperona) [inline composer.json:76]: You need to remove the require for monolog in the `require` above

File: inc/Dependencies/Psr/Log/Test/DummyTest.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<?php
```
Comment (remyperona) [inline inc/Dependencies/Psr/Log/Test/DummyTest.php]: Need to remove the tests classes

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
- Original issue isnot reproducible on test environment
- Fresh install  (using zip or Git) / update using zip working fine on Nginx
- Fresh install / update using zip on Apache working fine 
But on the test site (newlabs, mega) , the advanced cache isnot re-written when updating WPR

**Comment by vmanthos** [comment]
@CrochetFeve0251 When updating to `3.15` `rocket_generate_advanced_cache_file()` doesn't run as the following conditional evaluates to `true`:
https://github.com/wp-media/wp-rocket/blob/91dd642bc002f67c2c91957749210fc6c599c22d/inc/Engine/Cache/AdminSubscriber.php#L178

Shouldn't this be:
```PHP
                if ( version_compare( $old_version, '3.15', '>=' ) ) {
```
in order for the `advanced-cache.php` to be regenerated when updating from versions prior to 3.15?

**Review by remyperona — APPROVED** [review]

**Review by Mai-Saad — APPROVED** [review]
Working as expected with the following notes:
1- after update, the advanced cache won't be updated until visiting settings page.
2- Random warning :
disable debug at ticket then refresh our settings while file is imported and wpr updated
`[14-Jul-2023 06:22:28 UTC] PHP Warning:  touch(): Unable to create file /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/cache/wp-rocket/new.rocketlabsqa.ovh/home-parallax/.mobile-active because No such file or directory in /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/includes/class-wp-filesystem-direct.php on line 529`
3- Random error:
ollback after update while all settings on
```
[14-Jul-2023 06:30:39 UTC] PHP Fatal error:  Uncaught Error: Class "WP_Rocket\Plugin" not found in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/main.php:38
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(310): rocket_init()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(517): WP_Hook->do_action()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-settings.php(493): do_action()
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-config.php(119): require_once('...')
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-load.php(50): require_once('...')
#6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-cron.php(46): require_once('...')
#7 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/main.php on line 38
[14-Jul-2023 06:30:39 UTC] PHP Fatal error:  Uncaught Error: Class "WP_Rocket\Dependencies\Minify\JS" not found in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php:200
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php(130): WP_Rocket\Engine\Media\Lazyload\Subscriber->minify_script()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php(115): WP_Rocket\Engine\Media\Lazyload\Subscriber->add_inline_script()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(310): WP_Rocket\Engine\Media\Lazyload\Subscriber->insert_lazyload_script()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters()
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(517): WP_Hook->do_action()
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/general-template.php(3069): do_action()
#6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/themes/astra/footer.php(33): wp_footer()
#7 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/template.php(783): require_once('...')
#8 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/template.php(718): load_template()
#9 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/general-template.php(92): locate_template()
#10 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/themes/astra/page.php(44): get_footer()
#11 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/template-loader.php(106): include('...')
#12 /var/www/new.rocketlabsqa.ovh/htdocs/wp-blog-header.php(19): require_once('...')
#13 /var/www/new.rocketlabsqa.ovh/htdocs/index.php(17): require('...')
#14 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php on line 200
```
[testrail-report-504.pdf](https://github.com/wp-media/wp-rocket/files/12047725/testrail-report-504.pdf)

**Pull request merged**